### PR TITLE
Include `context` in the test `metadata.yaml` file

### LINKED ISSUE #2613: Accessing `context` within the test
An example use-case, I want `rpminspect` to skip `metadata` when `initiator == packit` because the copr metadata always fails. This would work with `adjust`, but in more complicated cases, I would like a more general approach that I can use at runtime.

- I have tried using the `$@`, but this i not expanded within `environment`:
  ```yaml
  environment:
    CONTEXT_INITIATOR: $@{initiator}
  ```
- I have tried looking for any other source where `context` are stored at runtime, but there doesn't seem to be any

Some potential solutions:
- Expand all `environment` nodes. This would be tricky if there is recursive dependence
- Add `context` to `TMT_TEST_METADATA`
- Export `TMT_CONTEXT_*` variables by default

### PR DESCRIPTION
**Pull request by LecrisUT**
Pull Request Checklist

* [x] implement the feature
* [x] extend the test coverage

Closes #2613

### CODE REVIEW

File: tmt/steps/execute/__init__.py [hunk 1]
```diff
@@ -410,9 +410,12 @@ def prepare_tests(self, guest: Guest, logger: tmt.log.Logger) -> list[TestInvoca
             invocation = TestInvocation(phase=self, test=test, guest=guest, logger=logger)
             invocations.append(invocation)
 
+            # Exported metadata is the test's metadata along with other variables like the context
+            test_metadata = test._metadata
```
Comment (happz) [inline tmt/steps/execute/__init__.py:414]: `.copy()` is needed, otherwise, the line below would be the same as `test._metadata['context'] = ...`

Comment (LecrisUT) [reply tmt/steps/execute/__init__.py]: I thought that the moment the variable is altered it becomes a local variable, but maybe `__setitem__` does not count for that. Anyway better to be safe

**Comment by LecrisUT** [comment]
Missing:
- Test metadata schema. Would be nice to have this somewhere.
- Documentation about TMT_TEST_METADATA. Doing a search I only found the environment variable.

File: tmt/steps/execute/__init__.py (hunk 1 above)
Comment (happz) [reply tmt/steps/execute/__init__.py]: Nope. Python variables aren’t variables, just labels. There is a dictionary somewhere in memory, pointed to by `test._metadata`, after the assignment it’s still the same dict, but now `test_metadata` points to it too. No implicit `copy()`.

**Comment by happz** [comment]
> Can `/plans/features/steps/execute ` be executed manually?

Yes. I do this often: with active development environment (`hatch -e dev shell`), `tmt -c trigger=commit -vv run -a provision -h local plan -n '^/plans/features/core$'`, for example. Or you can run a single test, `cd tests/core/adjust; ./test.sh`.

> Probably the failures are due to `test._metadata` or `test._fmf_context` being empty. The former should never be the case since `_update_metadata` should always populate `name`, and for thr latter I saw it uses a `FmfContext` constructor, so I'm not sure how it became like that.

There is definitely one issue I can see in tests, instead of direct assignment, `test_metadata["context"] = test._fmf_context`, we need to save context serialized, suitable for beign stored in a YAML structure, no special objects, as it crosses boundary between our code and the rest of the world, therefore `.to_spec()` is needed: `test_metadata["context"] = test._fmf_context.to_spec()`. This kills vast majority of tests, sorry about that.

**Comment by lukaszachy** [comment]
> Can `/plans/features/steps/execute ` be executed manually?

 `tmt run -avvv plans -n /plans/features/steps/execute`, but you should make sure you run with modified tmt. 
Thus either run in hatch/venv or create rpm and install it.

However easier to debug is usually pick one failed test and run it directly. Or run the tmt command from such test directly. There are too many layers which (IMHO) make debugging more painful than it should be.

In this case there is no need to run tests, as `cannot represent an object: {}` happens regularly and even this simple plan hits it:
```yaml
execute:
    how: tmt
    script: echo   
provision:
    how: local
```

**Comment by lukaszachy** [comment]
> * Test metadata schema. Would be nice to have this somewhere.

How about creating dir next to current step schemas, e.g. `tmt/schemas/variables` and document TMT_TEST_METADATA, TMT_TOPOLOGY_YAML (not sure it makes sense to have schema for TMT_TOPOLOGY_BASH)?

> * Documentation about `TMT_TEST_METADATA`. Doing a search I only found the environment variable. Where should the general documentation for that go?

Currently https://tmt.readthedocs.io/en/stable/overview.html#variables is the only place. How about creating new section in  https://tmt.readthedocs.io/en/stable/examples.html#tests to showcase how this envar can be used? Later we can add other variable examples.

All of above is nice to have though.

**Comment by happz** [comment]
> > * Test metadata schema. Would be nice to have this somewhere.

Isn't test metadata covered by tmt/schemas/test.yaml? IIUIC, it's the test metadata as fmf provided it, `self.node.data.copy()`.

> How about creating dir next to current step schemas, e.g. `tmt/schemas/variables` and document TMT_TEST_METADATA, TMT_TOPOLOGY_YAML (not sure it makes sense to have schema for TMT_TOPOLOGY_BASH)?

We definitely lack schema for the topology YAML, not sure we can have one for the Bash version.

**Comment by LecrisUT** [comment]
> therefore `.to_spec()` is needed:

Thanks, that indeed did the trick.

Now I've tried to run `^/tests/execute/metadata$` (thanks @happz for the tip with `hatch`), but surprisingly the `context` field is nowhere to be found, not even as an empty `{}` or `null`.

**Comment by LecrisUT** [comment]
> Isn't test metadata covered by tmt/schemas/test.yaml? IIUIC, it's the test metadata as fmf provided it, self.node.data.copy().

As far as i see it's the full expanded schema + extras like `context`, `extra-summary`, etc. One thing different is that in this case, all of the fields must be populated, so the schema test should be more robust on that point.

**Comment by LecrisUT** [comment]
The only leftover issue for this one is that it seems that `test._fmf_context` did not inherit the plan's context :thinking:. Probably that's not the appropriate field to use, or the metadata is written too early.

**Comment by lukaszachy** [comment]
We have some tips on development (including the hatch one) at https://tmt.readthedocs.io/en/stable/contribute.html#develop

**Comment by LecrisUT** [comment]
Things are a bit more fishy. I've added `context` to the test itself (although `context` is not in the test spec), and that is not picked up either. I looking at `test.tree._fmf_context`, this one has the context, but only the plan's. Should we make `test._fmf_context` initialize/point to `test.tree._fmf_context`?

**Comment by happz** [comment]
> Things are a bit more fishy. I've added `context` to the test itself (although `context` is not in the test spec), and that is not picked up either. Looking at `test.tree._fmf_context`, this one has the context, but only the plan's. Should we make `test._fmf_context` initialize/point to `test.tree._fmf_context`? My concern is that will miss the cli context

Hmm, test's fmf context might be a red herring :/ Can you try `self.step.plan._fmf_context` instead? It's the one we add to results after all, https://github.com/teemtee/tmt/blob/main/tmt/steps/execute/__init__.py#L557

**Comment by LecrisUT** [comment]
Ok, seems either `self.step.plan._fmf_context` and `test.tree._fmf_context` both work, including when using cli variable. Now the only issue is the normalization. All context seem to be normalized to array. Is that intended? What example is there where the context is used as an array? That's a bit unintuitive, but I guess it's workable.

**Comment by lukaszachy** [comment]
> Ok, seems either `self.step.plan._fmf_context` and `test.tree._fmf_context` both work, including when using cli variable. Now the only issue is the normalization. All context seem to be normalized to array. Is that intended? What example is there where the context is used as an array? That's a bit unintuitive, but I guess it's workable.

Dimension values being an array is implementation detail of fmf.context.Context. Each dimension can have multiple values

**Comment by LecrisUT** [comment]
> Dimension values being an array is implementation detail of fmf.context.Context. Each dimension can have multiple values

Oh, is it like [here](https://fmf.readthedocs.io/en/stable/context.html#dimensions) where `python3-3.8.5-5.fc32` -> `[python3, 3, 8, 5, 5, fc32]`. Documenting that would be tricky.

Anyway, the current version should work.

**Review by lukaszachy — APPROVED** [review]
@LecrisUT Do you wish to touch the documentatino/schema? 
If not we could include this into current (1.32) release.

**Comment by LecrisUT** [comment]
> @LecrisUT Do you wish to touch the documentatino/schema? If not we could include this into current (1.32) release.

Probably not this round. Would need some experimentation to see how we can reuse the test schema but make all of the fields required and sanitized.

**Comment by lukaszachy** [comment]
> Oh, is it like [here](https://fmf.readthedocs.io/en/stable/context.html#dimensions) where `python3-3.8.5-5.fc32` -> `[python3, 3, 8, 5, 5, fc32]`. Documenting that would be tricky.

Not really.  By the array I meant that context isn't `Dict[str, str]`. Sorry for the confusion.

The split into version parts is just example to understand how version compare in context matching works and is implementation detail.

Current way of saving context is correct - `Dict[str, List[str]]` and it should stay it that way

**Comment by LecrisUT** [comment]
> Current way of saving context is correct - `Dict[str, List[str]]` and it should stay it that way

Ok, but for working on the documentation, what should the user expect to be populated in that list? Because on the input side they provide it as a `dict[str, str]`

**Comment by happz** [comment]
> > Current way of saving context is correct - `Dict[str, List[str]]` and it should stay it that way
> 
> Ok, but for working on the documentation, what should the user expect to be populated in that list? Because on the input side they provide it as a `dict[str, str]`

Hm, that's probably unclear, but the value of a dimension is essentially a list of items. It looks like a string, but tmt 1. supports list in fmf and 2. expands the CLI value by splitting with `,`. These should be accepted by tmt, and represent a set of values of one dimension:

```
context:
  collection:
    - gcc-toolset-13
    - llvm-toolset-17

-c collection=gcc-toolset-13,llvm-toolset-17
```

So it *may* be a list, although I guess it rarely makes sense for the most common dimensions like `arch` or `trigger`.

What @lukaszachy shared is an example of how each of the values would be parsed for the purpose of evaluating `when` expressions. E.g. `distro: centos-8.3.0` would be split into "name" `centos`, and "version" components, `8`, `3`, `0`. But this split happens in fmf library when checking `when`, it's not how fmf context is represented in tmt and `FmfContext` class. There it's a simple `dict[str, list[str]`, and that's exactly what you get in the YAML:

```yaml
distro:
  - centos-8.3.0
collection:
  - gcc-toolset-13
  - llvm-toolset-17
```

**Comment by lukaszachy** [comment]
> Ok, but for working on the documentation, what should the user expect to be populated in that list? Because on the input side they provide it as a `dict[str, str]`

In tmt we have bunch of 'str' -> 'list[str]' expansions. 'component', 'contact', 'require' .... All we say is ` Must be a string or a list of strings.` So tmt accepts both and normalizes it to list[str]

**Comment by LecrisUT** [comment]
Thanks for the clarification @happz  @lukaszachy

**Review by happz — APPROVED** [review]

**Review by psss — APPROVED** [review]
Thanks for implementing this!

**Pull request merged**